File: C/5/5.c
#include <stdio.h>

#define MAXSTR 1000

/**
 * @brief Copies at most n chars from string s to t.
 *
 * @param s Source string
 * @param t Target string
 * @param n Number of characters
 */
void pstrncpy(char *s, char *t, int n)
{
    int i = 0;

    while (i++ < n && (*t++ = *s++))
        ;
}

/**
 * @brief Copies at most n chars from string s to the end of t.
 *
 * @param s Source string
 * @param t Target string
 * @param n Number of characters
 */
void pstrncat(char *s, char *t, int n)
{
    int i = 0;

    /* while *t is not '\0', go to the next char. */
    while (*t && t++)
        ;

    /**
     * That looks dangerous, we don't check for t's size.
     * But that's how they do it in the book, so we copy 's'
     * to the end of 't' while *s is not '\0'.
     */
    while ((*t++ = *s++) && i++ < n)
        ;

}

/**
 * @brief Compares at most n chars from string s and t.
 *
 * @param s     Source string
 * @param t     Target string
 * @param n     Number of characters to be matched
 *
 * @return int  1 if they match, 0 otherwise
 */
int pstrncmp(char *s, char *t, int n)
{
    int i = 0;

    /* Check if characters match */
    while ((*s++ == *t++) && ++i != n)
        ;

    /* If i == n all chars are matched */
    if (i == n)
        return 1;

    return 0;
}

/**
 * @brief Implementing a strcat function.
 */
main ()
{
    char t1[MAXSTR] = "The first string is";
    char s1[] = " IN NEED OF COMPLETION!";

    char t2[MAXSTR] = "abcdef678910";
    char s2[MAXSTR] = "12345";
    char s3[MAXSTR] = "123456789";

    printf("\"%s\"\t(pstrncat)=>\t", t1);
    pstrncat(s1, t1, 24);
    printf("\"%s\"\n", t1);

    printf("\"%s\"\t(pstrncpy)=>\t", t2);
    pstrncpy(s2, t2, 5);
    printf("\"%s\"\n", t2);

    printf("\"%s\"\tpstrncmp\t\"%s\"=>\t%d\n", s2, s3, pstrncmp(s2, s3, 5));
    printf("\"%s\"\tpstrncmp\t\"%s\"=>\t%d\n", t1, s2, pstrncmp(t1, s2, 5));
}

File: rustlang/5/5.rs
/// @brief Copies at most n chars from string s to t.
///
/// @param s Source string
/// @param t Target string
/// @param n Number of characters
fn pstrncpy(s: &str, t: &mut String, n: i32)
{
    let mut i: i32 = 0;

    for _c in s.chars() {
        if i >= n {
            break;
        }

        t.remove(i as usize);
        t.insert(i as usize, _c);
        i += 1;
    }
}

/// Copies at most n chars from string s to the end of t.
///
/// @param s Source string
/// @param t Target string
/// @p
fn pstrncat(s: &str, t: &mut String, n: i32)
{
    let mut i: i32 = 0;

    for _c in s.chars() {

        if i >= n {
            break;
        }

        t.push(_c);
        i += 1;
    }
}

/// Compares at most n chars from string s and t.
///
/// @param s      Source string
/// @param t      Target string
/// @param n      Number of characters to be matched
///
/// @return bool  True if they match, false otherwise
fn pstrncmp(s: &str, t: &str, n: i32) -> bool
{
    let mut i: i32 = 0;

    let mut s_iter = s.chars();
    let mut t_iter = t.chars();

    // Check if characters match
    while (s_iter.next() == t_iter.next()) && i != n {
        i += 1;
    }

    // If i == n all chars are matched
    if i == n {
        return true;
    }

    false
}

/// Implementing a strcat function.
fn main ()
{
    let mut t1 = String::from("The first string is");
    let s1 = String::from(" IN NEED OF COMPLETION!");

    let mut t2 = String::from("abcdef678910");
    let s2 = String::from("12345");
    let s3 = String::from("123456789");

    print!("\"{}\"\t(pstrncat)=>\t", t1);
    pstrncat(&s1, &mut t1, 24);
    print!("\"{}\"\n", t1);

    print!("\"{}\"\t(pstrncpy)=>\t", t2);
    pstrncpy(&s2, &mut t2, 5);
    print!("\"{}\"\n", t2);

    print!("\"{}\"\tpstrncmp\t\"{}\"=>\t{}\n", s2, s3, pstrncmp(&s2, &s3, 5));
    print!("\"{}\"\tpstrncmp\t\"{}\"=>\t{}\n", t1, s2, pstrncmp(&t1, &s2, 5));
}
